lib/auth/github.go

```diff
@@ -84,6 +91,118 @@ func (a *Server) upsertGithubConnector(ctx context.Context, connector types.Gith
+type HTTPRequester interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// checkGithubOrgSSOSupport returns an error if any of the Github
+// organizations specified in this connector use external SSO. SSO
+// is a Teleport Enterprise feature that should not be allowed for OSS.
+// If userTeams is not nil, only organizations that are both specified
+// in conn and in userTeams will be checked. If client is nil a
+// net/http.Client will be used.
+func checkGithubOrgSSOSupport(ctx context.Context, conn types.GithubConnector, userTeams []teamResponse, orgCache *utils.FnCache, client HTTPRequester) error {
+	version := modules.GetModules().BuildType()
+	if version == modules.BuildEnterprise {
+		return nil
+	}
+
+	orgs := make(map[string]struct{})
+	addOrg := func(org string) {
+		if len(userTeams) != 0 {
+			// Only check organizations that the user is a member of and
+			// that are specified in this auth connector
+			for _, team := range userTeams {
+				if org == team.Org.Login {
+					orgs[org] = struct{}{}
+				}
+			}
+		} else {
+			orgs[org] = struct{}{}
+		}
+	}
+
+	// Check each organization only once
+	for _, mapping := range conn.GetTeamsToLogins() {
+		addOrg(mapping.Organization)
+	}
+	for _, mapping := range conn.GetTeamsToRoles() {
+		addOrg(mapping.Organization)
+	}
+
+	if client == nil {
+		transport, err := defaults.Transport()
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		client = &http.Client{
+			Transport: transport,
+		}
+	}
+
+	checkOrg := func(org string) (bool, error) {
+		// A Github organization will have a "sso" page reachable if it
+		// supports external SSO. There doesn't seem to be any way to get this
+		// information from the Github REST API without being an owner of the
+		// Github organization, so check if this exists instead.
+		ssoURL := fmt.Sprintf("%s/%s/%s", githubOrgsURL, url.PathEscape(org), "sso")
+
+		const retries int = 3
+		var resp *http.Response
+		for i := 0; i < retries; i++ {
+			ctx, cancel := context.WithTimeout(ctx, defaults.HTTPRequestTimeout)
+			defer cancel()
+
+			req, err := http.NewRequestWithContext(ctx, http.MethodGet, ssoURL, nil)
+			if err != nil {
+				return false, trace.Wrap(err)
+			}
+
+			var urlErr *url.Error
+			resp, err = client.Do(req)
+			if err == nil {
+				break
+			} else if errors.As(err, &urlErr) && urlErr.Timeout() {
+				if i == retries-1 {
+					// The connection timed out a couple of times in a row,
+					// stop trying and return the error.
+					return false, trace.ConnectionProblem(err, "Timed out trying to reach GitHub to check for organization external SSO.")
+				}
+				// Connection timed out, try to make the request again
+				continue
+			}
+			// Unknown error, don't try making any more requests
+			return false, trace.Wrap(err, "Unknown error trying to reach GitHub to check for organization external SSO")
+		}
+		resp.Body.Close()
+
+		if resp.StatusCode == http.StatusOK {
```

Comment:
What would be the status code if SSO is not enabled? 404? Might be worth mentioning in a comment.

Reply:
Done